onnxruntime/core/framework/allocation_planner.cc

```diff
@@ -238,6 +241,36 @@ class PlannerImpl {
     return false;
   }
 
+  bool FindReusableGraphInput(const onnxruntime::Node& node, int output_arg_num, MLValueIndex* reusable_input) {
+    // skip if this is main graph
+    if (parent_node_) {
+      // only check a single layer. ie. an Identity/Dropout node connecting graph's input and output
+      auto& graph_inputs = graph_viewer_.GetInputs();
+      auto& graph_outputs = graph_viewer_.GetOutputs();
+
+      if (FindReusableInput(node, output_arg_num, reusable_input)) {
+        auto& arg = ml_value_info_.at(Buffer(*reusable_input)).p_def_site;
+        auto it = std::find(graph_inputs.cbegin(), graph_inputs.cend(), arg);
+        if (it != graph_inputs.end()) {
+          auto graph_input_index = std::distance(graph_inputs.cbegin(), it);
+          ORT_ENFORCE(graph_input_index >= 0 && static_cast<size_t>(graph_input_index) < graph_inputs.size());
+          auto graph_output_index = std::distance(graph_outputs.cbegin(),
+                                                  std::find(graph_outputs.cbegin(), graph_outputs.cend(), node.OutputDefs()[output_arg_num]));
+          ORT_ENFORCE(graph_output_index >= 0 && static_cast<size_t>(graph_output_index) < graph_outputs.size());
+
+          if (parent_node_->OpType() == "Loop") {
+            // the matching carried parameters
+            return graph_input_index >= 2 && graph_input_index == graph_output_index + 1;
+          } else {
+            // TODO: other operators like Scan
+          }
```

Comment:
It would be better if this information is provided in a more generic manner so that we're not hardcoding special casing for a list of operators here. 

e.g. something like the kernel def could be expanded and we lookup that information based on OpType.

Reply:
I am fully agree with the point. However I think we need a design agreement before I make a change since this change may apply to the core type `KernelOp`.